packages/loader/container-loader/src/catchUpMonitor.ts

```diff
@@ -0,0 +1,101 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDisposable, IEvent } from "@fluidframework/common-definitions";
+import { assert, TypedEventEmitter } from "@fluidframework/common-utils";
+import { IDeltaManager } from "@fluidframework/container-definitions";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+/** @see ICatchUpMonitor for usage */
+type CaughtUpListener = (hasCheckpointSequenceNumber: boolean) => void;
+
+/** @see ICatchUpMonitor for usage */
+export interface ICatchUpMonitorEvents extends IEvent {
+    (event: "caughtUp", listener: CaughtUpListener): void;
+}
+
+/** Monitor that emits an event when a Container has caught up to a given point in the op stream */
+export interface ICatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents>, IDisposable { }
+
+/**
+ * Monitors a Container's DeltaManager, notifying listeners when all ops have been processed
+ * that were known at the time the monitor was created.
+ */
+export class CatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents> implements ICatchUpMonitor {
+    private readonly hasCheckpointSequenceNumber: boolean;
+    private readonly targetSeqNumber: number;
+    private caughtUp: boolean = false;
+
+    private readonly opHandler = (message: Pick<ISequencedDocumentMessage, "sequenceNumber">) => {
+        if (!this.caughtUp && message.sequenceNumber >= this.targetSeqNumber) {
+            this.caughtUp = true;
+            this.emit("caughtUp", this.hasCheckpointSequenceNumber);
+        }
+    };
+
+    /**
+     * Create the CatchUpMonitor, setting the target sequence number to wait for based on DeltaManager's current state.
+     */
+    constructor(
+        private readonly deltaManager: IDeltaManager<any, any>,
+    ) {
+        super();
+
+        this.hasCheckpointSequenceNumber = this.deltaManager.hasCheckpointSequenceNumber;
+        this.targetSeqNumber = this.deltaManager.lastKnownSeqNumber;
+
+        assert(this.targetSeqNumber >= this.deltaManager.lastSequenceNumber,
+            "Cannot wait for seqNumber below last processed sequence number");
+
+        this.deltaManager.on("op", this.opHandler);
+
+        // Simulate the last processed op to set caughtUp in case we already are
+        this.opHandler({ sequenceNumber: this.deltaManager.lastSequenceNumber });
+
+        // If a listener is added after we are already caught up, notify that new listener immediately
+        this.on("newListener", (event: string, listener) => {
```

Comment:
It feels like instead of using events, it's better to use a property that exposes a promise. I.e. something like
```ts
public caughtUp: Promise<number>;
```
Promise seems to be better because this event files once, and past "firing" reflected on new listeners - that's exactly what promise does.

Reply:
That is a great point.  I don't remember why I switched to event-based to begin with, but I bet it would work better with Promise.  That said, I don't have time to try it out now, and also leaving it will reduce merge conflicts with your follow-on refactor.